File: InterviewPrepare/MaximumSumCircularSubarray.java
/**
 * LeetCode 918 Maximum Sum Circular Subarray
 */
class Solution {
    public int maxSubarraySumCircular(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int sumAll = 0;
        int maxSum = nums[0];
        int minSum = nums[0];
        int curMax = 0 ,curMin = 0;
        for (int i = 0; i < nums.length; i++) {
            curMax = Math.max(curMax + nums[i],nums[i]);
            curMin = Math.min(curMin + nums[i],nums[i]);
            maxSum = Math.max(maxSum,curMax);
            minSum = Math.min(minSum,curMin);
            sumAll += nums[i];
        }
        return maxSum > 0 ? Math.max(maxSum,sumAll - minSum) : maxSum;
    }
}

File: MaximumSumCircularSubarray.go
func maxSubarraySumCircular(nums []int) int {
	n := len(nums)
	leftMax := make([]int,n)

	leftMax[0] = nums[0]
	leftSum,pre,res := nums[0],nums[0],nums[0]
	for i := 1;i < n;i++ {
		pre = max(pre + nums[i],nums[i])
		res = max(res,pre)
		leftSum += nums[i]
		leftMax[i] = max(leftMax[i - 1],leftSum)
	}

	rightSum := 0
	for i := n - 1;i > 0;i-- {
		rightSum += nums[i]
		res = max(res,rightSum + leftMax[i - 1])
	}
	return res
}

func max(a int,b int) int {
	if (a > b) {
		return a
	}
	return b
}

func min(a int,b int) int {
	if (a < b) {
		return a
	}
	return b
}
